src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/TypeCast.cs

```diff
@@ -42,119 +42,147 @@ internal enum AssignmentVariation
+            Debug.Assert(mt != pTargetType, "The check for the trivial cases should be inlined by the JIT");
+
+            if (!mt->IsCanonical)
             {
-                // arrays can be cast to System.Object
-                if (WellKnownEETypes.IsSystemObject(pTargetType))
-                {
-                    return obj;
-                }
+                // Arrays should be the only non-canonical types that can exist on GC heap
+                Debug.Assert(mt->IsArray);
 
-                // arrays can be cast to System.Array
-                if (WellKnownEETypes.IsSystemArray(pTargetType))
-                {
-                    return obj;
-                }
+                // arrays can be cast to System.Object or System.Array
+                if (WellKnownEETypes.IsValidArrayBaseType(pTargetType))
+                    goto done;
 
-                return null;
+                // They don't cast to any other class
+                goto fail;
             }
 
-
-            // walk the type hierarchy looking for a match
-            while (true)
+            for (; ; )
             {
-                pObjType = pObjType->NonArrayBaseType;
-                if (pObjType == null)
-                {
-                    return null;
-                }
+                mt = mt->NonArrayBaseType;
+                if (mt == pTargetType)
+                    goto done;
 
-                if (pObjType == pTargetType)
-                {
-                    return obj;
-                }
-            }
-        }
+                if (mt == null)
+                    break;
 
-        [RuntimeExport("RhTypeCast_CheckCastClass")]
-        public static unsafe object CheckCastClass(MethodTable* pTargetEEType, object obj)
-        {
-            // a null value can be cast to anything
-            if (obj == null)
-                return null;
+                mt = mt->NonArrayBaseType;
+                if (mt == pTargetType)
+                    goto done;
 
-            object result = IsInstanceOfClass(pTargetEEType, obj);
+                if (mt == null)
+                    break;
 
-            if (result == null)
-            {
-                // Throw the invalid cast exception defined by the classlib, using the input MethodTable*
-                // to find the correct classlib.
+                mt = mt->NonArrayBaseType;
+                if (mt == pTargetType)
+                    goto done;
 
-                throw pTargetEEType->GetClasslibException(ExceptionIDs.InvalidCast);
+                if (mt == null)
+                    break;
+
+                mt = mt->NonArrayBaseType;
+                if (mt == pTargetType)
+                    goto done;
+
+                if (mt == null)
+                    break;
             }
 
-            return result;
+            goto fail;
+
+        done:
+            return obj;
+
+        fail:
+            throw pTargetType->GetClasslibException(ExceptionIDs.InvalidCast);
```

Comment:
This can benefit from tailcalling the slow path or tailcalling throw helper. Tailcalling should to make the method frame smaller - it should save a few instructions in prolog/epilog.

Reply:
Pushed an update that should make it tailcallable but it needs someone with RyuJIT whisperer skills because it's still not a tail call (notice frame setup and a `call` at the end):

```
124:         [RuntimeExport("RhTypeCast_CheckCastClassSpecial")]
   125:         private static unsafe object CheckCastClassSpecial(MethodTable* pTargetType, object obj)
   126:         {
   127:             Debug.Assert(!pTargetType->IsParameterizedType, "CheckCastClass called with parameterized MethodTable");
   128:             Debug.Assert(!pTargetType->IsFunctionPointerType, "CheckCastClass called with function pointer MethodTable");
   129:             Debug.Assert(!pTargetType->IsInterface, "CheckCastClass called with interface MethodTable");
   130:             Debug.Assert(!pTargetType->HasGenericVariance, "CheckCastClass with variant MethodTable");
   131: 
   132:             MethodTable* mt = obj.GetMethodTable();
00007FF6047592C0 48 83 EC 28          sub         rsp,28h  
00007FF6047592C4 4C 8B 02             mov         r8,qword ptr [rdx]  
   133:             Debug.Assert(mt != pTargetType, "The check for the trivial cases should be inlined by the JIT");
   134: 
   135:             if (!mt->IsCanonical)
00007FF6047592C7 41 F7 00 00 00 03 00 test        dword ptr [r8],30000h  
00007FF6047592CE 74 26                je          S_P_CoreLib_System_Runtime_TypeCast__CheckCastClassSpecial+36h (07FF6047592F6h)  
   136:             {
   137:                 // Arrays should be the only non-canonical types that can exist on GC heap
   138:                 Debug.Assert(mt->IsArray);
   139: 
   140:                 // arrays can be cast to System.Object or System.Array
   141:                 if (WellKnownEETypes.IsValidArrayBaseType(pTargetType))
00007FF6047592D0 44 8B 01             mov         r8d,dword ptr [rcx]  
00007FF6047592D3 41 C1 E8 1A          shr         r8d,1Ah  
00007FF6047592D7 41 83 E0 1F          and         r8d,1Fh  
00007FF6047592DB 41 83 F8 16          cmp         r8d,16h  
00007FF6047592DF 74 0D                je          S_P_CoreLib_System_Runtime_TypeCast__CheckCastClassSpecial+2Eh (07FF6047592EEh)  
00007FF6047592E1 41 83 F8 14          cmp         r8d,14h  
00007FF6047592E5 75 47                jne         S_P_CoreLib_System_Runtime_TypeCast__CheckCastClassSpecial+6Eh (07FF60475932Eh)  
00007FF6047592E7 48 83 79 08 00       cmp         qword ptr [rcx+8],0  
00007FF6047592EC 75 40                jne         S_P_CoreLib_System_Runtime_TypeCast__CheckCastClassSpecial+6Eh (07FF60475932Eh)  
   176:                     break;
   177:             }
   178: 
   179:             goto fail;
   180: 
   181:         done:
   182:             return obj;
00007FF6047592EE 48 8B C2             mov         rax,rdx  
   183: 
   184:         fail:
   185:             return ThrowInvalidCastException(pTargetType);
00007FF6047592F1 48 83 C4 28          add         rsp,28h  
00007FF6047592F5 C3                   ret  
   142:                     goto done;
   143: 
   144:                 // They don't cast to any other class
   145:                 goto fail;
   146:             }
   147: 
   148:             for (; ; )
   149:             {
   150:                 mt = mt->NonArrayBaseType;
00007FF6047592F6 4D 8B 40 08          mov         r8,qword ptr [r8+8]  
   151:                 if (mt == pTargetType)
00007FF6047592FA 4C 3B C1             cmp         r8,rcx  
00007FF6047592FD 74 EF                je          S_P_CoreLib_System_Runtime_TypeCast__CheckCastClassSpecial+2Eh (07FF6047592EEh)  
   152:                     goto done;
   153: 
   154:                 if (mt == null)
00007FF6047592FF 4D 85 C0             test        r8,r8  
00007FF604759302 74 2A                je          S_P_CoreLib_System_Runtime_TypeCast__CheckCastClassSpecial+6Eh (07FF60475932Eh)  
   155:                     break;
   156: 
   157:                 mt = mt->NonArrayBaseType;
00007FF604759304 4D 8B 40 08          mov         r8,qword ptr [r8+8]  
   158:                 if (mt == pTargetType)
00007FF604759308 4C 3B C1             cmp         r8,rcx  
00007FF60475930B 74 E1                je          S_P_CoreLib_System_Runtime_TypeCast__CheckCastClassSpecial+2Eh (07FF6047592EEh)  
   159:                     goto done;
   160: 
   161:                 if (mt == null)
00007FF60475930D 4D 85 C0             test        r8,r8  
00007FF604759310 74 1C                je          S_P_CoreLib_System_Runtime_TypeCast__CheckCastClassSpecial+6Eh (07FF60475932Eh)  
   162:                     break;
   163: 
   164:                 mt = mt->NonArrayBaseType;
00007FF604759312 4D 8B 40 08          mov         r8,qword ptr [r8+8]  
   165:                 if (mt == pTargetType)
00007FF604759316 4C 3B C1             cmp         r8,rcx  
00007FF604759319 74 D3                je          S_P_CoreLib_System_Runtime_TypeCast__CheckCastClassSpecial+2Eh (07FF6047592EEh)  
   166:                     goto done;
   167: 
   168:                 if (mt == null)
00007FF60475931B 4D 85 C0             test        r8,r8  
00007FF60475931E 74 0E                je          S_P_CoreLib_System_Runtime_TypeCast__CheckCastClassSpecial+6Eh (07FF60475932Eh)  
   169:                     break;
   170: 
   171:                 mt = mt->NonArrayBaseType;
00007FF604759320 4D 8B 40 08          mov         r8,qword ptr [r8+8]  
   172:                 if (mt == pTargetType)
00007FF604759324 4C 3B C1             cmp         r8,rcx  
00007FF604759327 74 C5                je          S_P_CoreLib_System_Runtime_TypeCast__CheckCastClassSpecial+2Eh (07FF6047592EEh)  
   173:                     goto done;
   174: 
   175:                 if (mt == null)
00007FF604759329 4D 85 C0             test        r8,r8  
00007FF60475932C 75 C8                jne         S_P_CoreLib_System_Runtime_TypeCast__CheckCastClassSpecial+36h (07FF6047592F6h)  
   183: 
   184:         fail:
   185:             return ThrowInvalidCastException(pTargetType);
00007FF60475932E E8 DD 14 00 00       call        S_P_CoreLib_System_Runtime_TypeCast__ThrowInvalidCastException (07FF60475A810h)  
00007FF604759333 CC                   int         3  
```

Reply:
The throw helper needs to be marked as `[MethodImpl(MethodImplOptions.NoInlining)]` in order for the JIT to tailcall it.

When it is not marked as NoInlining, the JIT will try to inline it, makes an observation that the call never returns, and the tailcall optimization then gets disabled here: https://github.com/dotnet/runtime/blob/a951e996134660303b26941ab5d1359d8350be35/src/coreclr/jit/morph.cpp#L5827-L5831